File: 2023/python/day_02.py
# https://adventofcode.com/2023/day/2


INPUT_FILE: str = "2023/data/day_02.txt"
BUDGET_LIMITS: dict[str, int] = {
    "red": 12,
    "green": 13,
    "blue": 14,
}


class Game:
    id: int
    rounds: list[dict[str, int]]

    def __init__(self, id: int, rounds: list[dict[str, int]]):
        self.id = id
        self.rounds = rounds

    def __repr__(self) -> str:
        return f"Game(id={self.id}, rounds={self.rounds})"


def parse_round(round_string: str) -> dict[str, int]:
    round_parts = [x.strip() for x in round_string.split(",")]
    return {color: int(number) for number, color in [x.split(" ") for x in round_parts]}


def get_input(input_file: str) -> list[Game]:
    with open(input_file, "r") as f:
        lines = [x.strip() for x in f.readlines()]

    games = []
    for line in lines:
        game_part = line.split(":")
        game_id = int(game_part[0].split(" ")[1])
        round_strings = [x.strip() for x in game_part[1].split(";")]
        games.append(Game(id=game_id, rounds=[parse_round(x) for x in round_strings]))

    return games


def is_valid(game: Game, budget: int) -> bool:
    for game_round in game.rounds:
        for color, number in game_round.items():
            if number > budget[color]:
                return False
    return True


def get_game_power(game: Game) -> int:
    current_game = {"red": 0, "green": 0, "blue": 0}
    for round in game.rounds:
        for color, number in round.items():
            current_game[color] = max(current_game[color], number)
    result = 1
    for _, number in current_game.items():
        result *= number
    return result


def solve_day_02_part_1(input_file: str = INPUT_FILE):
    result = 0
    games = get_input(input_file)
    for game in games:
        if is_valid(game, BUDGET_LIMITS):
            result += game.id
    return result


def solve_day_02_part_2(input_file: str = INPUT_FILE):
    result = 0
    games = get_input(input_file)
    for game in games:
        result += get_game_power(game)
    return result


if __name__ == "__main__":
    print(f"Solution for Part 1: {solve_day_02_part_1(INPUT_FILE)}")
    print(f"Solution for Part 2: {solve_day_02_part_2(INPUT_FILE)}")

File: 2023/rust/src/day_02.rs
// https://adventofcode.com/2023/day/2

use std::fs::File;
use std::io::Read;
use std::path::Path;

#[derive(Clone)]
struct GameRound {
    red: i32,
    green: i32,
    blue: i32,
}

struct Game {
    id: i32,
    rounds: Vec<GameRound>,
}

const BUDGET_LIMITS: GameRound = GameRound {
    red: 12,
    green: 13,
    blue: 14,
};

fn parse_game_round(game_round: &str) -> GameRound {
    let mut red: i32 = 0;
    let mut green: i32 = 0;
    let mut blue: i32 = 0;
    for round_part in game_round.trim().split(", ").collect::<Vec<&str>>() {
        let round_part_split = round_part.split(" ").collect::<Vec<&str>>();
        match round_part_split[1] {
            "red" => {
                red = match round_part_split[0].parse::<i32>() {
                    Ok(value) => value,
                    Err(_) => panic!("Invalid number {}", round_part_split[0]),
                }
            }
            "green" => {
                green = match round_part_split[0].parse::<i32>() {
                    Ok(value) => value,
                    Err(_) => panic!("Invalid number {}", round_part_split[0]),
                }
            }
            "blue" => {
                blue = match round_part_split[0].parse::<i32>() {
                    Ok(value) => value,
                    Err(_) => panic!("Invalid number {}", round_part_split[0]),
                }
            }
            _ => panic!("Invalid color {}", round_part_split[1]),
        }
    }
    GameRound {
        red: red,
        green: green,
        blue: blue,
    }
}

fn read_file_contents(input_fname: impl AsRef<Path>) -> Vec<Game> {
    let file_path = input_fname.as_ref();
    let mut input = File::open(file_path).expect("Failed to open file");
    let mut contents = String::new();
    let _ = input
        .read_to_string(&mut contents)
        .expect("Failed to read file");

    let mut games: Vec<Game> = vec![];
    for line in contents.lines() {
        let mut game: Game = Game {
            id: 0,
            rounds: vec![],
        };

        let game_rounds_split = line.split(": ").collect::<Vec<&str>>();
        game.id = game_rounds_split[0].split(" ").collect::<Vec<&str>>()[1]
            .parse::<i32>()
            .unwrap();
        for game_round in game_rounds_split[1].split(";") {
            game.rounds.push(parse_game_round(game_round).clone());
        }
        games.push(game);
    }
    games
}

fn is_game_valid(game: &Game, budget_limits: GameRound) -> bool {
    for game_round in &game.rounds {
        if game_round.red > budget_limits.red
            || game_round.green > budget_limits.green
            || game_round.blue > budget_limits.blue
        {
            return false;
        }
    }
    true
}

fn get_game_power(game: &Game) -> i32 {
    let mut game_max: GameRound = GameRound {
        red: 0,
        green: 0,
        blue: 0,
    };

    for game_round in &game.rounds {
        game_max.red = game_max.red.max(game_round.red);
        game_max.green = game_max.green.max(game_round.green);
        game_max.blue = game_max.blue.max(game_round.blue);
    }
    game_max.red * game_max.green * game_max.blue
}

pub fn solve_day_02_part_1(input_fname: impl AsRef<Path>) -> i32 {
    let mut result: i32 = 0;
    let games = read_file_contents(input_fname);
    for game in games {
        if is_game_valid(&game, BUDGET_LIMITS) {
            result += game.id;
        }
    }
    result
}

pub fn solve_day_02_part_2(input_fname: impl AsRef<Path>) -> i32 {
    let mut result: i32 = 0;
    let games = read_file_contents(input_fname);
    for game in games {
        result += get_game_power(&game);
    }
    result
}
